File: src/dataset/javascript/forth_exercism-ThomasZumsteg/claude/ClaudeSonnet4_forth_v3.js
function n_args(n, func) {
  return function(stack) {
    if (stack.length < n)
      throw new Error("Stack empty");
    const args = stack.splice(-n);
    const result = func(...args);
    if (result.length > 0) {
      stack.push(...result);
    }
  }
}

export default class Forth {
  constructor() {
    this.stack = [];
    this.env = new Map();
    this.builtin = new Map([
      ["+", n_args(2, (a, b) => [a + b])],
      ["-", n_args(2, (a, b) => [a - b])],
      ["*", n_args(2, (a, b) => [a * b])],
      ["/", n_args(2, (a, b) => {
        if (b === 0)
          throw new Error("Division by zero");
        return [Math.floor(a / b)];
      })],
      ["dup", n_args(1, a => [a, a])],
      ["drop", n_args(1, () => [])],
      ["swap", n_args(2, (a, b) => [b, a])],
      ["over", n_args(2, (a, b) => [a, b, a])]
    ]);
    this.numberRegex = /^-?\d+$/;
  }

  evaluate(line) {
    if (line.startsWith(': ') && line.endsWith(' ;')) {
      const tokens = line.slice(2, -2).split(" ");
      const head = tokens[0].toLowerCase();
      if (this.numberRegex.test(head))
        throw new Error("Invalid definition");
      this.env.set(head, tokens.slice(1));
      return;
    }

    const tokens = line.split(" ");
    let i = 0;
    
    while (i < tokens.length) {
      const token = tokens[i].toLowerCase();
      
      if (this.numberRegex.test(token)) {
        this.stack.push(parseInt(token, 10));
      } else if (this.env.has(token)) {
        tokens.splice(i, 1, ...this.env.get(token));
        continue;
      } else if (this.builtin.has(token)) {
        this.builtin.get(token)(this.stack);
      } else {
        throw new Error('Unknown command');
      }
      i++;
    }
  }
}

File: src/dataset/python/forth_exercism-ThomasZumsteg/claude/ClaudeSonnet4_forth_v3.py
class StackUnderflowError(Exception):
    pass


keywords = {
    '+': (2, lambda a, b: (a + b,)),
    '-': (2, lambda a, b: (a - b,)),
    '*': (2, lambda a, b: (a * b,)),
    '/': (2, lambda a, b: (a // b,)),
    'swap': (2, lambda a, b: (b, a)),
    'over': (2, lambda a, b: (a, b, a)),
    'dup': (1, lambda x: (x, x)),
    'drop': (1, lambda x: tuple()),
}


def evaluate(input_data):
    tokens = []
    for line in input_data:
        tokens.extend(line.split())
    
    stack = []
    env = {}
    i = 0
    
    while i < len(tokens):
        token = tokens[i].lower()
        
        if token == ':':
            i += 1
            keyword = tokens[i].lower()
            if keyword.isdigit():
                raise ValueError("Variables cannot be numbers: " + keyword)
            
            i += 1
            definition = []
            while tokens[i] != ';':
                definition.append(tokens[i])
                i += 1
            env[keyword] = definition
            
        elif token in env:
            tokens[i:i+1] = env[token]
            continue
            
        elif token in keywords:
            n_args, func = keywords[token]
            if len(stack) < n_args:
                raise StackUnderflowError('Cannot apply ' + token)
            
            if n_args == 1:
                arg = stack.pop()
                result = func(arg)
            else:
                args = [stack.pop() for _ in range(n_args)]
                args.reverse()
                result = func(*args)
            
            stack.extend(result)
            
        else:
            stack.append(int(token))
        
        i += 1
    
    return stack